### PR TITLE
Ensure security log is closed on LifeCycle.shutdown

### PR DESCRIPTION
**Pull request by craigtaverner**
There is a bug that only happens on windows, and can be seen in the StoreCopyClientTest when deleting log files, the security log cannot be deleted due to a lock.

There are many differences between the implementation of SecurityLog and the other logs. One that appears very relevant to this case is the fact that the debug.log is in LogService which implements LifeCycle, and when the shutdown() method is called, the LogService closes the OutputStream to the debug.log (See StoreLogService.java:195).

So I've implemented LifeCycle on SecurityLog and added calls to init and shutdown from the existing LifeCycle methods on the enterprise auth manager. We could have added SecurityLog to the lifecycle directly, as is done with debug.log in PlatformModule, but the edition differences make that messy, so it was easier to add these calls in the MultiRealmAuthManager where the security log is created and managed directly.

changelog: Fixed problem on windows with security log not being closed before backup is run, by tying it to the lifecycle shutdown

### CODE REVIEW

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/MultiRealmAuthManager.java [hunk 1]
```diff
@@ -187,6 +188,7 @@ public void shutdown() throws Throwable
                 ((RealmLifecycle) realm).shutdown();
             }
         }
+        securityLog.shutdown();
```
Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/MultiRealmAuthManager.java]: I noted that `EnterpriseAuthManagerFactory` has a piece of defensive code that can either accept an externally injected log or create a new one. 
`private SecurityLog getSecurityLog( Log allegedSecurityLog )`

If we get the log injected we should be careful about how we manage its life cycle.

File: enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/SecurityLog.java [hunk 2]
```diff
@@ -38,35 +39,31 @@
...
 
     public SecurityLog( Config config, FileSystemAbstraction fileSystem, Executor executor ) throws IOException
     {
-        this( createLog( config, fileSystem, executor ) );
-    }
-
-    public SecurityLog( Log log )
-    {
-        inner = log;
-    }
```
Comment (henriknyman) [inline enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/SecurityLog.java:53]: I believe this alternative constructor is still used by both tests and `EnterpriseAuthManagerFactory.getSecurityLog`.

**Review by henriknyman — CHANGES_REQUESTED** [review]

File: enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/SecurityLog.java [hunk 3]
```diff
@@ -38,35 +39,31 @@
 import static org.neo4j.helpers.Strings.escape;
 import static org.neo4j.kernel.impl.enterprise.configuration.EnterpriseEditionSettings.security_log_filename;
 
-public class SecurityLog implements Log
+public class SecurityLog implements Log, Lifecycle
```
Comment (MishaDemianenko) [inline enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/SecurityLog.java]: Why not to extend LifecycleAdapter instead for example and just override `shutdown`?

File: enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/SecurityLog.java (hunk 2 above)
Comment (craigtaverner) [reply enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/SecurityLog.java:53]: I reviewed the usage and the only non-test place it is used it gets from EditionModel.authManagerLog, which is always either the real security log, or a null log. So this means we're back to two cases, a closable log or a non-closable one, and that is dealt with in the condition inside the shutdown method.

I would prefer a better approach, but the use of two constructors for setting up SecurityLog is a pre-existing design issue I did not want to resolve in this PR, which focuses only on adding the close() call to the real security log (and not to test or null-logs).

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/MultiRealmAuthManager.java (hunk 1 above)
Comment (craigtaverner) [reply enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/MultiRealmAuthManager.java]: Agreed, but as described in my other response below, I see the issue with the two SecurityLog constructors as orthogonal to this fix. Injected logs are not closable or support lifecycle, so we cannot interact with them at all.

File: enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/SecurityLog.java (hunk 3 above)
Comment (craigtaverner) [reply enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/SecurityLog.java]: Good idea.

File: enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/SecurityLog.java (hunk 2 above)
Comment (henriknyman) [reply enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/SecurityLog.java:53]: Sure, I mostly meant that it wouldn't compile without it. But I see now that you have added it again below.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/MultiRealmAuthManager.java (hunk 1 above)
Comment (henriknyman) [reply enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/MultiRealmAuthManager.java]: The `if ( this.rotatingSupplier != null )` accomplishes that

**Review by henriknyman — APPROVED** [review]
